### DISCUSSION TITLE
Euphony build 및 test 실행

### DISCUSSION THREAD

**Discussion by yeonns** [euphony-io/euphony#45] [Knowledge Share]
## Euphony Build
Android Studio에서 ```[Build - Make Project]```를 통해 build할 수 있습니다.
❗ Euphony는 native code를 포함하기 때문에, build하기 전에 ```[Tools - SDK Manager - SDK Tools]```에서 NDK, CMake, Android SDK Platform-Tools를 설치해주세요.  

### ⚠️ 주의사항: ADB 경로 (by @YoungSeokHong)
Euphony build시에 adb가 ```${ANDROID_SDK_ROOT}/platform-tools```에 위치해야 정상동작합니다:
![image](https://user-images.githubusercontent.com/10149398/133630858-cb73afb5-e1ca-4875-bb51-baf45958380a.png)

💡 Android Studio의 SDK Manager를 통해 ```Android SDK Platform-Tools```를 설치했다면, 자동으로 해당 경로에 adb가 위치하게 되니, 신경쓰지 않으셔도 됩니다.

### Build 성공 확인
Build가 성공한다면 아래와 같이 ```BUILD SUCCESSFUL``` 메세지를 확인할 수 있습니다:  
![image](https://user-images.githubusercontent.com/10149398/133432991-da92748c-80c3-478e-8606-c0800612f115.png)

### FAQ) Build가 실패합니다
- NDK와 CMake가 설치되어 있는지 확인해보세요
- adb가 SDK Location밑의 ```platform-tools``` 내에 위치하는지 확인해보세요 (SDK Location은 Android Studio내의 ```[Tools - SDK Manager]```에서 확인할 수 있습니다)
- terminal/powershell에서 ```adb devices``` 명령어가 동작하는지 확인해보세요
- ```euphony/.cxx``` 디렉토리를 삭제한 후, 다시 build 해보세요

## Euphony Native Test
Euphony의 native code와 test code가 작성된 위치는 다음과 같습니다:  
- Native Code: euphony/src/main/cpp/core/source
- Native Test Code: euphony/src/main/cpp/tests

현재 Euphony의 native test는 **device**(android 기기 혹은 AVD)가 연결되어 있을 때, build를 하면 실행되도록 ```euphony/src/main/cpp/tests/CMakeLists.txt```에 작성되어있습니다. 
따라서 연결된 **device**가 없을 경우, native test는 실행되지 않습니다.

만약에 device 없이 build한 후에, device를 연결시키고 다시 build를 하면 build error가 발생합니다. 이처럼 device 연결이 변경된 경우에는 ```euphony/.cxx``` 디렉토리를 전부 삭제한 후에, 다시 build를 해줘야 error가 발생하지 않습니다.
참고로 ```.cxx``` 디렉토리를 삭제하면, project clean 후 새로 build하는 것과 동일하기 때문에, build 시간이 오래 걸립니다.

### Euphony Native Test Fail Result
test code가 실패하도록 코드를 수정한 후, build해봅시다.  
Test file: euphony/src/main/cpp/tests/base16Test.cpp
```cpp
TestParamType(std::vector<u_int8_t>{ 0x61 }, "621"),
```
Build output을 통해 test code fail을 확인할 수 있습니다:  
![image](https://user-images.githubusercontent.com/10149398/133434236-db445954-4f8a-4bab-a55d-bd766ccc91e7.png)

### FAQ) Test가 실패인데도, build가 성공합니다 (by @kmsaidev)
- ```euphony/.cxx``` 디렉토리를 삭제하고, 다시 build 해보세요
- 가상기기(AVD)를 사용해서 build중이라면, 가상기기가 제대로 연결되어있는지 ```adb devices```를 통해 확인해주세요

**Comment by YoungSeokHong** [comment]
## 저하고 같은 경우에 있으신 분들 있으실 것 같아서 comment 올립니다!

native test code 에서 fail이 뜨는 것을 실습해보기 위해서 ADB를 설치하고 디바이스(Galaxy Note 8)를 연결한 뒤,
euphony/src/main/cpp/tests/base16Test.cpp 파일의 32번 라인을 수정하고 build를 해봤습니다.
```cpp
TestParamType(std::vector<u_int8_t>{ 0x61 }, "621"),
```
그런데 build가 오류 하나 없이 잘 돌아갔습니다.
![image](https://user-images.githubusercontent.com/80624082/133606494-9a8a3b66-29cb-4a56-8d04-1ab775007286.png)

ABD가 제대로 작동하지 않아서 디바이스가 연결되지 않았나해서 cmd창에서 ```adb devices```명령어로 확인을 해봤더니 디바이스 연결에는 이상이 없었습니다.
![image](https://user-images.githubusercontent.com/80624082/133606828-c418cc0f-af21-4be5-a909-cec1779610f3.png)

그래서 euphony/src/main/cpp/tests/CMakeList.txt파일을 확인하던 중 36번 라인에서 ADB를 불러오는 경로가 잘못된 것 같아서
![image](https://user-images.githubusercontent.com/80624082/133607054-3a989747-dea3-4bc3-a18d-eb5546ec7b5e.png)
경로를 제가 ADB를 설치할 때 platform-tools폴더를 저장해뒀던 경로인 ```C:/dev/platform-tools```로 변경했습니다.
```cmake
find_program(ADB NAMES adb PATHS C:/dev/platform-tools)
```
그 후 다시 build를 실행하니 반가운(?) 에러메세지를 볼 수 있었습니다!
![image](https://user-images.githubusercontent.com/80624082/133607597-3b1c5b8a-79eb-4f30-8a77-c667171d1f1b.png)

PS. CMake를 다뤄보는 건 처음이라 아직까지 ```${ANDROID_SDK_ROOT}```가 무엇인지 이해하지 못하고 있습니다 ㅠㅠ 제대로 된 해결 방안을 제시한건지도 모르겠네요😂 우선 저와 같은 경우에 계신 분이 있으시면 저처럼 임시방편으로 ADB경로를 수정해주시던지, ADB설치 위치와 환경변수 설정을 변경해주셔야 할 것 같습니다! ```${ANDROID_SDK_ROOT}```에 대해서 아시는 분은 comment부탁드립니다!

**Reply by YoungSeokHong** [reply]
찾아보니 ```ANDROID_SDK_ROOT```는 시스템 환경변수이고 제가 아직 ```ANDROID_SDK_ROOT```를 만든적이 없어서 못 찾았던 것 같습니다! 우선은 ```ANDROID_SDK_ROOT```환경변수를 Adroid SDK Location을 참고해서 만들어 줬고 
![image](https://user-images.githubusercontent.com/80624082/133613907-7ecbee30-c617-4f4b-b3ad-b2f91c0477d6.png)

환경변수를 만들어 준 뒤 platform-tools 디렉토리를 해당 경로로 옮겨줬더니 ADB가 제대로 작동했습니다.

환경변수를 사용 하는 것을 응용해서 ```ANDROID_PLATFORM_TOOLS``` 시스템 환경변수를 platform-tools 디렉토리가 있는 경로로 만든 뒤,
euphony/src/main/cpp/tests/CMakeList.txt 파일의 36번 라인을 아래와 같이 바꿔줬더니
```cmake
find_program(ADB NAMES adb PATHS ${ANDROID_PLATFORM_TOOLS})
```
이또한 제대로 원하던 결과를 얻을 수 있었습니다.

**Reply by yeonns** [reply]
@YoungSeokHong 완벽한 해결이네요! 😮 제 개발환경에서는 해당 경로에 이미 adb가 존재해서 이런 이슈가 없었기 때문에 이부분을 놓쳤는데, 꼭 필요했던 부분을 작성해주셔서 감사합니다 👍 

저는 아래와 같은 환경이고, 영석님이 변경해주셨던 36번라인 변경없이도 build가 잘 이루어집니다!
- ```ANDROID_SDK_ROOT``` 및 ```ANDROID_PLATFORM_TOOLS``` 가 환경변수로 저장X
- SDK Location: C:\Users\user\AppData\Local\Android\Sdk
- ADB Location: C:\Users\user\AppData\Local\Android\Sdk\platform-tools (cmake에 정의된 것처럼 SDK Location의 바로 하위 directory에 위치함)

말씀해주신 cmake의 36번째라인처럼 로컬 컴퓨터의 SDK ROOT와 platform-tools의 위치를 맞춰주기만해도 환경변수 설정 없이도 될거같은데 혹시 시도해보실수 있을까요??
```cpp
find_program(ADB NAMES adb PATHS ${ANDROID_SDK_ROOT}/platform-tools) # Verified to be working on Linux.
```

**Reply by designe** [reply]
ㅎㅎ 참고로 ${ANDROID_SDK_ROOT} 는 build.gradle에 정의되어 있으니 참고해주세요 :)
https://github.com/euphony-io/euphony/blob/54966925eb8a78d72e47eb31acd03c270f4c9694/euphony/build.gradle#L17

기본적으로 Euphony Native 빌드를 위한 cmake는 build.gradle에 의해 실행이 됩니다.
여기 라인에 보시면 build.gradle에서 cmake의 arguguments로 -DANDROID_SDK_ROOT 라고 되어 있는데 이게 Define ANDROID_SDK_ROOT의 의미입니다. android.getSdkDirectory().getAbsolutePath()를 ANDROID_SDK_ROOT라는 변수에 저장하는건데요.

음, 영석님의 경우, android.getSdkDirectory().getAbsolutePath()를 gradle에서 인식하지 못해서 발생한 문제로 보입니다.
gradle 버전에 따라서 이는 다르게 동작할 수 있는 부분이긴 합니다만! 현재 세팅된 gradle에서는 정상동작 해야하는데 인식되지 못했다는 것은... 패치가 예상치 못한 부분에서 나올수도 있을 것 같은 느낌이네요!!!

**Reply by designe** [reply]
@YoungSeokHong gtest 동작 성공하신걸 추카드립니다!! 캬캬 👍

**Reply by YoungSeokHong** [reply]
정말 감사합니다! ㅠㅠ 이제 퍼즐이 맞춰진 느낌입니다. Euphony의 경우 ANDROID_SDK_ROOT가 이미 build.gradle에 정의되어있었군요!
제가 처음에 오류가 났었던 이유가 platform-tools 디렉토리를 SDK의 경로가 아닌 처음에 말씀드렸던 C:/dev에 넣어놨어서 그런 것 같습니다. 말씀하신 대로 해보기 위해서 제가 설정해둔 ```ANDROID_SDK_ROOT```환경변수를 지우고 SDK 경로에 platform-tools 디렉토리를 넣고 돌려봤더니 정상적으로 작동했습니다! 

build.gradle에서 -D로 변수를 정의해 주는 것도 오늘 처음 배워갑니다. 감사합니다!

**Reply by designe** [reply]
@YoungSeokHong 근데 궁금한 점이 있다면 Android SDK 폴더에는 platform-tools 디렉토리가 원래 있어야 하는데 영석님은 왜 없었는지 궁금하네요? 혹시 platform-tools 폴더를 수동으로 옮기셨던건가요?!

**Reply by yeonns** [reply]
@designe android studio 내의 sdk manager를 통해 platform-tools를 설치했다면 괜찮았을텐데, 제가 adb 설치를 다른 경로에서 하게끔 가이드를 작성해서 해당 위치에 없었던거같아요! 가이드 수정하겠습니다 ㅎㅎ

**Reply by designe** [reply]
> @designe android studio 내의 sdk manager를 통해 platform-tools를 설치했다면 괜찮았을텐데, 제가 adb 설치를 다른 경로에서 하게끔 가이드를 작성해서 해당 위치에 없었던거같아요! 가이드 수정하겠습니다 ㅎㅎ

@yeonns 아하 확인했습니다. sdk manager가 아닌 수동으로 platform-tools를 쓴 경우군요!! 오키

@yeonns @YoungSeokHong platform-tools를 수동으로 설치하는 경우에 대해서 대응이 안된 것으로 보이네요.
드물지만 충분히 있을 수 있는 상황으로 보입니다. build.gradle에서 Sdk 폴더에 platform-tools 여부를 확인한 후, 없는 경우 platform-tools가 없다고 에러 메세지를 날려주는 패치를 한번 만들어보는건 어떨까요? ㅎㅎ

**Reply by yeonns** [reply]
@designe 좋은 생각이에요! 해당 경로에 adb가 없을 경우 에러를 발생시키고, SDK Manager를 통해 platform-tools를 설치하라는 메세지까지 출력하도록 하면 좋을 거 같습니다 😄 
@YoungSeokHong CMakeLists 작성하는 걸 공부하신 다음에 커밋도전!?ㅋㅋ

**Reply by YoungSeokHong** [reply]
저도 처음에는 정신없이 이것저것 해보느라 정확히 기억하지는 못하겠습니다. ㅠㅠ
제가 다시 처음 환경부터 차례대로 해봤을 때 예상 되는 것은
1. 처음에 디바이스를 연결하지 않은 상태에서 build를 하고
2. 디바이스를 연결하고나서 .cxx디렉토리을 실수로 지우지 않고 다시 build를 해보고 오류메시지가 뜨지 않자
3. ADB가 잘못되어있다고 생각하고 재설치를 하는 과정에서 platform-tools를 전혀 엉뚱한 경로에 저장을 해서 헤맨것 같습니다
결국 처음에 .cxx디렉토리을 지우는 것을 생각하지 않고 있다가 이것저것 뜯어보다가 그런 것 같습니다.ㅠㅠ

만약에 저처럼 .cxx디렉토리를 지우는 것을 깜빡하지 않고, ADB도 재설치하지 않는 일반적인 환경이라면 이러한 문제는 발생하지 않을 것 같습니다!

수렁으로 빠지느라 이것저것 많이 알게 된 것 같습니다. 결국 .cxx디렉토리를 통으로 지우는 방법이 멘토님께서 미션에서 말씀하신 코드를 고치지 않고 해결하는 방법인 것 같고, 코드를 고쳐서 해결하는 방법을 다시 열심히 찾아봐야겠습니다!! 늦은 시간인데 계속해서 신경써주셔서 감사합니다ㅠㅠ

**Comment by kmsaidev** [comment]
## 저도 저와 같은 경우의 분들이 계실까봐 comment 올립니다.
저도 euphony/src/main/cpp/tests/base16Test.cpp 파일 32번째 라인을 수정 후 build를 해보았는데 테스트 실패가 뜨지 않고 빌드에 성공했습니다.</br>
```cpp
TestParamType(std::vector<u_int8_t>{ 0x61 }, "621"),
```
뭐가 원인인지 찾다가 가상기기가 제대로 켜져있지 않은 걸 알게 되었습니다.</br>
![20210916234948](https://user-images.githubusercontent.com/88221290/133634233-92d0baa9-1837-4251-a2bb-ff05686f8c91.png)</br>
혹시 저처럼 가상기기를 사용하여 native test code를 실행하시는 분들이 계시다면</br>
![20210916233938](https://user-images.githubusercontent.com/88221290/133633642-e711426d-b602-4afc-ada5-eaebb81df330.png)</br>
꼭 cmd창에 adb devices 명령어를 입력 후 가상기기가 잘 켜져있는지 확인 후 빌드하시면 좋을 것 같습니다. </br>
저도 이 문제를 해결하고 나니 test code fail을 확인할 수 있었습니다!</br>
![20210916235409](https://user-images.githubusercontent.com/88221290/133635201-98223de0-4b84-4036-a84d-573441a1e3d8.png)

**Reply by yeonns** [reply]
@YoungSeokHong @kmsaidev 로컬에서 build가 제대로 되지 않는 경험을 공유해주시고, 해결방법까지 알려주셔서 정말 감사합니다 😄 덕분에 build 및 test 실행방법 discussion이 점점 더 정교해지고있습니다!! (제가 작성시에 놓친 부분이 많다는 걸 새삼 깨닫게 되네요 😆)

**Reply by designe** [reply]
가상기기 혹은 단말의 상태가 connect 되어 있지 않은 경우에는 유닛테스트는 진행되지 않고, 빌드만 진행되도록 구성이 되어 있습니다 :)
그 이유는 native 코드의 테스트는 실제 안드로이드 에뮬 혹은 단말에서만 진행이 가능하기 때문인데요. 
다들 잘하고 계시는군요ㅎㅎ

**Comment by judemin** [comment]
## adb 명령어가 cmd에서 작동하지 않을 경우에 대한 해결책입니다

![image](https://user-images.githubusercontent.com/20484492/133888484-a5d93367-7ef3-47c0-93d2-421be96502c3.png)
분명 위와 같이 SDK Manager에서 Platform-Tools를 설치했는데도 불구하고 해당 discussion에서 다른 분이 쓰시는 것 처럼 adb와 adb devices를 똑같이 cmd에서 실행시켰을 때 명령어를 찾을 수 없다는 메세지가 나왔었습니다.

이런 경우에는 설치 폴더에 정상적으로 설치되어 있는것을 확인하고 간단하게 PATH환경 변수에 adb가 들어있는 폴더의 위치를 추가하면 됩니다. 윈도우에서 폴더의 위치는 아래와 같습니다.

```
C:\Users\<사용자이름>\AppData\Local\Android\Sdk\platform-tools
```

환경변수는 윈도우 제어판에서 [시스템]->[고급시스템 설정]->[환경 변수]에서 시스템 변수중 PATH에 추가하면 됩니다.
![image](https://user-images.githubusercontent.com/20484492/133888654-fb6bbc48-04be-4f62-94a0-e9187ea1694d.png)

**Reply by psh2849** [reply]
저도 같은 이유로 좀 헤맸었는데 좋은 정보인 것 같습니다!
